### ISSUE TITLE
add stealable keyword

### ISSUE THREAD

**Issue by abast** [dask/distributed#847]
Inspired by this [stackoverflow question](http://stackoverflow.com/questions/41965253/repeated-task-execution-using-the-distributed-dask-scheduler/41965766#41965766), I want to suggest adding a `idempotent` keyword somewhere in the workflow, which in the case `idempotent = False` switches off work stealing. I would suggest to make it a keyword of a delayed object, in this case a workflow that expects tasks to be run only once could look like this:

```
c = distributed.Client()
delayed(mytask, idempotent = False).compute(get = c.get)
```

**Comment by mrocklin** [comment]
Adding features like this is expensive, not just the development time to add them, but the extra inertia that it adds to the project as we move forward.  Every feature like this makes development on the scheduler and reasoning about what's going on harder.  I would prefer to see if this issue comes up more before frequently committing to make a change like this.

Also, the term "idempotent" means "can be run multiple times without negative effect", so a different term would probably be desired like "stealable".

**Comment by abast** [comment]
Okay, I understand that. Generally it would be nice to have a good way to react if a task runs twice. Are there alternatives that are both convenient for the user and easy to implement?

Also I noticed that in case something does not work with the multiprocessing scheduler, the general advice always was "try the distributed scheduler", which perfectly worked as a drop-in replacement. Due to work-stealing, this might not work anymore in a number of cases.

**Comment by mrocklin** [comment]
Okay, I understand that. Generally it would be nice to have a good way to react if a task runs twice. Are there alternatives that are both convenient for the user and easy to implement?

I can't think of anything at the moment.  My inclination would be to try to resolve this by making tasks idempotent (can be run multiple times without bad effect) and not depending on side effects like local files.  Or, if you require local files, tie their cleanup to cleanup within the python process using `__del__` or the `atexit` module.

Generally speaking Dask doesn't play nicely with code that depends on side effects.  My recommendations are usually going to be about avoiding or mitigating the use of side effects.

**Comment by abast** [comment]
The files are not local, but on a NAS. These files are basically the desired result of the computation. I don't want to clean them up, but I have to make sure that the same simulation does not occur twice.

Not relying on side effects would mean that I have to send ~200GB of data through the network at the end of the simulation process, which could easily be stored during that process. This not only adds a time consuming step at the end of the pipeline, it also raises the questions when this data can be freed from the memory.

**Comment by abast** [comment]
If a task is stolen, is it guaranteed that both tasks get the time to finish? Or might it happen that the slower worker receives e.g. a SIGTERM by the scheduler?

**Comment by mrocklin** [comment]
If a task has already started running in a thread then we don't have a good
way to stop it without killing the worker.  Both tasks will finish.
However ideally the stolen task hasn't yet begun at the time of stealing.

**Comment by abast** [comment]
Perfect, thanks. Basically a function that relies on side effects could implement a short check at the beginning to figure out if the respective result is already there. If the function decides to compute the result, the scheduler will not interfere. 

Maybe such a check (already calculated or not) could take advantage of https://github.com/dask/distributed/issues/872?

**Comment by mrocklin** [comment]
Yes.  Generally the solution to side effects and uncertainty from machine
failure, work stealing, etc. is to make functions idempotent as you
suggest.

Yes, I agree that some sort of centralized metadata store would probably be
a good tool for users to accomplish this.

**Comment by abast** [comment]
When will `my_delayed.compute(get = c.get)` return? Will it wait until **all** tasks that have been started are completed?

**Comment by mrocklin** [comment]
yes.  The collection.compute method is blocking.  The client.compute method
is non-blocking and returns a Future

**Comment by abast** [comment]
So I have to be more specific here. Regarding the discussion above (how to deal with the combination work stealing + side effects) I had the following special case in mind: 

- The scheduler sends the same job to two workers (which can happen because of work stealing). 
- The job relies on side effects but has a short check at the beginning, which checks if the actual computation has already been started and if so, it returns immediately. 
- Therefore the first call to that function will take long (hours, here the actual computation takes place), but the second call will quickly return (milliseconds, since the function does nothing more than check if the actual computation is already running).

For the scheduler it should look like the computation of that job is finished as soon as the second worker that gets the job returns some value. So I would expect, that the actual computation will continue to run on the cluster (because you said, the scheduler does not kill or interrupt a worker). But is it guaranteed, that the scheduler will wait for the "slowest job" before it returns?

**Comment by mrocklin** [comment]
From the scheduler's and client's perspective the task will finish once the first returns

**Comment by abast** [comment]
How can I block until the last job associated with the original delayed object has finished?

**Comment by mrocklin** [comment]
As far as I know there is no way to accomplish this today.  The
cluster-wide locking service would probably resolve your problem.  You
could maybe hack something up with channels?

**Comment by abast** [comment]
Okay, thanks. ... asking google about channels just gave me some django related links: https://pypi.python.org/pypi/channels .... this is probably not what you had in mind?

How do you currently manage writes with to_csv / to_hdf on the distributed scheduler? What happens if two processes try to write to the same file? How can you guarantee that there are no processes still writing files when the scheduler returns a value?

... maybe I can adapt something from your solutions.

**Comment by mrocklin** [comment]
We don't.  We always write to separate files

http://distributed.readthedocs.io/en/latest/channels.html
You can also append normal values, not just futures.  As of
https://github.com/dask/distributed/pull/858

**Comment by abast** [comment]
So if the task, that writes a single partition to a single csv file executes twice because of work stealing, we would get two separate files containing the same data?

**Comment by mrocklin** [comment]
Dask.dataframe.to_csv writes each partition to a different CSV file.  If it
happens twice then the second will overwrite the first.

**Comment by abast** [comment]
And if they write at the same time? Can I disable work stealing by changing the function `steal_time_ratio` in `stealing.py` such that it always returns `None, None`?

**Comment by mrocklin** [comment]
Sure.  You could also remove the WorkStealing extension when instantiating
the scheduler


    def __init__(self, center=None, loop=None,
                 delete_interval=500, synchronize_worker_interval=60000,
                 services=None, allowed_failures=ALLOWED_FAILURES,
                 extensions=[ChannelScheduler, PublishExtension,
WorkStealing],
                 validate=False, scheduler_file=None, **kwargs):

Or you could turn off the periodic callback

scheduler.extensions['stealing']._pc.stop()

**Comment by abast** [comment]
Great. Thanks for your support!

**Comment by abast** [comment]
Just for the record, to switch out work stealing from within a IPython session, this seems to work:

```
def switch_of_work_stealing(dask_scheduler=None):
    dask_scheduler.extensions['stealing']._pc.stop()

c = distributed.Client()
c.run_on_scheduler(switch_of_work_stealing)
```

@mrocklin: 
1. how can I switch the callback back on?
2. the scheduler seems to schedule a high number of tasks per worker, leading to a long processing stack per worker. In case of long running tasks and if work stealing is disabled, this is not optimal. Is there a way to change the scheduling depth?

**Comment by mrocklin** [comment]
> how can I switch the callback back on?

Search for `pc.start` in distributed/stealing.py


> the scheduler seems to schedule a high number of tasks per worker, leading to a long processing stack per worker. In case of long running tasks and if work stealing is disabled, this is not optimal. Is there a way to change the scheduling depth?

No.  All tasks are scheduled to some worker immediately.  Stealing is used to balance afterwards.

**Comment by abast** [comment]
Would the following do the trick?

```
def switch_of_work_stealing(dask_scheduler=None):
    dask_scheduler.extensions['stealing']._pc.start()
```

If so, I would put it on stack overflow.

**Comment by mrocklin** [comment]
I don't know.  Can tornado periodic callbacks be started after they've been
stopped?  You might need to try it out.

**Comment by abast** [comment]
It runs (does not give an error), but I am also not sure if it really runs properly. So I'll try it and if it works, post it on stackoverflow soon ("How to temporarily disable work stealing" or something). 

Usecase: This can be an important feature in science, where you might have some published reference implementation of something that is thoroughly tested by hand by some former PhD student but does not come with a test suite. In these cases, you try to avoid any changes to this code if possible, i.e. making it idempotent can become a major task. If work stealing is something that can be switched off easily, dask distributed becomes an option to parallelize such code.

**Comment by tonylampada** [comment]
I just asked [this question](http://stackoverflow.com/questions/44072816/dask-stopping-workers-during-execution-causes-completed-tasks-to-be-launched-t) on stackoverflow, and then found this issue. It seems my use case for dask-distributed is similar to @abast's.

If I understand correctly there's currently no way to accomplish that (tell dask to orchestrate distributed processing with no regard for state management). Is that correct?

**Comment by xwang777** [comment]
Looking at distributed/stealing.py move_task function, it seems that the scheduler process handles the move of  task from one worker to another, which in a distributed setting is not atomic. If instead the scheduler  only notifies a pair of workers to start a task move session between them, then the victim worker can atomically gives away tasks that are not running yet to the thief worker in the end-to-end communication. I'm not sure about the details of distributed, but you think this could be a solution?

**Comment by mrocklin** [comment]
Work stealing is now transactional.  It is still possible that a task will run more than once if a worker dies mid-stream, but this is fairly unlikely.  Closing this.

**Comment by abast** [comment]
This is awesome, thanks!
